refresh.template.py

```diff
@@ -212,36 +215,105 @@ def _get_cpp_command_for_files(compile_action: json):
         # Revert when https://github.com/clangd/clangd/issues/123 is solved, which would remove the need to emit headers, because clangd would take on that work.
         # If https://github.com/clangd/clangd/issues/681, we'd probably want to find a way to filter to one entry per platform.
         header_files = [h for h in header_files if h not in header_file_entries_written]
         header_file_entries_written.update(header_files)
 
         for file in itertools.chain(source_files, header_files):
-            sys.stdout.write(json.dumps({
-                'file': file,
-                'command': command,
-                'directory': bazel_workspace_dir
-            }, indent=2, check_circular=False))
-            sys.stdout.write(',')
+            yield {
+                "file": file,
+                "command": command,
+                # Bazel gotcha warning: If you were tempted to use `bazel info
+                # execution_root` as the build working directory for
+                # compile_commands...search ImplementationReadme.md to learn why
+                # that breaks.
+                "directory": os.fspath(directory),
+            }
+
+
+# Call with the same flags as `bazel build`, one target per call, followed by flags
+# Yields the entries to compile_commands.json
+def get_commands(directory: pathlib.Path, target: str, flags: str) -> str:
+    # Log clear completion messages
+    print(f"\033[0;34m>>> Analyzing commands used in {target}\033[0m")
+
+    # Queries Bazel's C-family compile actions, and runs them through our extract.py reformatter
+    # Aquery docs if you need em: https://docs.bazel.build/versions/master/aquery.html
+    # One bummer, not described in the docs, is that aquery filters over *all* actions for a given target, rather than just those that would be run by a build to produce a given output. This mostly isn't a problem, but can sometimes surface extra, unnecessary, misconfigured actions. Chris has emailed the authors to discuss and filed an issue so anyone reading this could track it: https://github.com/bazelbuild/bazel/issues/14156.
+    # We switched to jsonproto instead of proto because of https://github.com/bazelbuild/bazel/issues/13404. We could change back when fixed--reverting most of the commit that added this line and tweaking the build file to depend on the target in that issue. That said, it's kinda nice to be free of the dependency, unless (OPTIMNOTE) jsonproto becomes a performance bottleneck compated to binary protos.
+
+    cmd = [
+        "bazel",
+        "aquery",
+        f"mnemonic('(Objc|Cpp)Compile',deps({target}))",
+        "--ui_event_filters=-info",
+        "--noshow_progress",
+        "--output=jsonproto",
+    ] + shlex.split(flags)
+
+    completed = subprocess.run(
+        cmd,
+        cwd=os.fspath(directory),
+        encoding="utf-8",
+        errors="replace",
+        check=True,
+        stdout=subprocess.PIPE,
+    )
```

Comment:
Of the things here, I'm probably most concerned about what happens here when there are errors.

Ideally, we'd have the following behavior, I think:
- Still report errors from aquery to the user, ideally including warnings.
- Continue on when sub-targets fail.
- And shush the known/fine warnings parallel to how we previously did with sed in the shell script.
  -  Without this last one, we'll get lots of people unnecessarily worried and filing issues.

Perhaps the simplest, if not quite the best, would be printing stderr from a try block iff there was an error, and then continuing to the next command? And the best would be always printing a filtered stderr and catching errors with a try.

Either way I'd also love to preserve the note about the missing graph targets error message so it doesn't throw people for a loop.

Reply:
Done as a fixup.